publish.js

```diff
@@ -0,0 +1,287 @@
+        var packCmd = options.packCmd || defaults.packCmd;
+        publish.execSync(packCmd);
+    } else {
+        // publish to npm
+        var publishCmd = options.publishCmd || defaults.publishCmd;
+        publish.execSync(publishCmd);
+    }
+};
+
+/**
+ * Tags the specified version with the specified dist-tag
+ * If it is a test run, the tag command will be output to the console.
+ *
+ * @param isTest {Boolean} - indicates if this is a test run or not
+ * @param version {String} - a string indicating which version to tag
+ * @param tag {String} - the dist-tag to apply
+ * @param options {Object} - e.g. "npm dist-tag add ${packageName}@${version} ${tag}"
+ */
+publish.tag = function (isTest, packageName, version, tag, options) {
+    var cmdTemplate = options.distTagCmd || defaults.distTagCmd;
+    var cmdStr = es6Template(cmdTemplate, {
+        packageName: packageName,
+        version: version,
+        tag: tag
+    });
+    if (isTest) {
+        publish.log("tag command: " + cmdStr);
+    } else {
+        publish.execSync(cmdStr);
+    }
+};
+
+/**
+ * Restore the package.json file to the latest committed version.
+ *
+ * Used internally to reset version number changes in package.json
+ * @param packagePath {String} - the path to the package.json file to clean
+ * @param options {Object} - e.g. {"cleanCmd": "git checkout -- ${package}"}
+ */
+publish.clean = function (packagePath, options) {
+    var cmdTemplate = options.cleanCmd || defaults.cleanCmd;
+    var cmdStr = es6Template(cmdTemplate, {
+        "package": packagePath
+    });
+    publish.execSync(cmdStr);
+};
+
+/**
+ * Publishes a development build.
+ * This creates a release named after the version, but with the build stamp
+ * appended to the end. By default this will create a release with version
+ * X.x.x-prerelease.yyyymmddThhmmssZ.shortHash where X.x.x is sourced
+ * from the version number in the package.json file, -pre-release is from the
+ * devTag option (also applied as a tag to the release), and the build id
+ * (yyyymmddThhmmssZ.shortHash) is generated based on the latest commit.
+ *
+ * @param isTest {Boolean} - indicates if this is a test run
+ * @param options {Object} - see defaultOptions in package.json for possible values
+ */
+publish.dev = function (isTest, options) {
+    var opts = extend(true, {}, defaults, options);
+
+    // The package.json file of the top level package which is
+    // running this module.
+    var modulePkgPath = path.join(opts.moduleRoot || process.cwd(), "package.json");
+    var modulePkg = require(modulePkgPath);
+
+    // Ensure no uncommitted changes
+    publish.checkChanges(opts);
+
+    var devVersion = publish.getDevVersion(modulePkg.version, opts);
+
+    // set the version number
+    publish.setVersion(devVersion, opts);
+
+    publish.pubImpl(isTest, opts);
+    publish.tag(isTest, modulePkg.name, devVersion, opts.devTag, opts);
+
+    // cleanup changes
+    publish.clean(modulePkgPath, opts);
```

Comment:
Ok - I've determined that we can resolve the problem on cygwin by simply sending "package.json" here rather than modulePkgPath. The issue is that the set of absolute file paths which are dispensed by node don't agree with the ones accepted by git - but relative paths are fine. However, we do then have the problem that we just introduced, that the directory might not be the current working directory. In that case, what I suggest is that you issue 

https://nodejs.org/api/process.html#process_process_chdir_directory

to change the working directory to opts.moduleRoot before issuing clean, and then again to restore the original cwd.